app/src/main/java/com/example/music_app/data/repositories/LoginRepositoryImpl.kt

```diff
@@ -1,50 +1,93 @@
 import com.github.michaelbull.result.Err
 import com.github.michaelbull.result.Ok
+import com.github.michaelbull.result.Result
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.flow
 import kotlinx.coroutines.flow.flowOn
-import kotlinx.coroutines.flow.map
+import java.time.Instant
 import java.util.Base64
 
 private const val AUTH_HEADER_NAME = "Authorization"
 private const val AUTH_PROPERTY = "Basic"
 private const val TYPE_HEADER_NAME = "Content-Type"
 private const val TYPE_HEADER = "application/x-www-form-urlencoded"
-private const val KEY = "access_token"
+
+private const val TOKEN_KEY = "access_token"
+private const val TIME_KEY = "token_receiving_time"
+private const val REFRESH_TOKEN_KEY = "refresh_token"
+
+private const val AUTH_GRANT_TYPE = "authorization_code"
+private const val REFRESH_GRANT_TYPE = "refresh_token"
+
+private const val TOKEN_LIFETIME = 3600
 
 class LoginRepositoryImpl(
-    private val dataStoreManager: DataStoreManager
-): LoginRepository {
-    private val spotifyAPI = AuthService.getInstance()
+    private val dataStoreManager: DataStoreManager,
+    private val spotifyAPI: AuthService = AuthService.getInstance()
+) : LoginRepository {
 
     override suspend fun requestToken(code: String) = flow {
-        val token = spotifyAPI.getToken(createHeaders(), GRANT_TYPE, code, REDIRECT_URL)
-        emit(if (token.accessToken != null) {
-            saveToken(token.accessToken)
-            Ok(token)
-        } else Err(ResponseError()))
+        val token = spotifyAPI.getToken(createHeaders(), AUTH_GRANT_TYPE, code, REDIRECT_URL)
+        emit(
+            if (token.accessToken != null) {
+                saveToken(token)
+                Ok(token)
+            } else Err(ResponseError())
+        )
     }.flowOn(Dispatchers.IO)
 
-    override suspend fun saveToken(token: String) = dataStoreManager.saveString(token, KEY)
+    override suspend fun requestRefreshToken(): Flow<Result<TokenRefreshResponse, ResponseError>> =
+        flow {
+            val refreshToken = dataStoreManager.getString(REFRESH_TOKEN_KEY)
+            val token =
+                spotifyAPI.getRefreshToken(createHeaders(), REFRESH_GRANT_TYPE, refreshToken)
+            emit(
+                if (token.accessToken != null) {
+                    saveRefreshToken(token)
+                    Ok(token)
+                } else Err(ResponseError())
+            )
+        }
 
-    override fun isAuthorized(): Flow<Boolean> =
-        dataStoreManager.getString(KEY).map(String::isNotEmpty)
+    override suspend fun isAuthorized(): Boolean =
+        dataStoreManager.getString(TOKEN_KEY).isNotEmpty()
+
+    override suspend fun isOutdated(): Boolean {
+        val recTime = Instant.parse(dataStoreManager.getString(TIME_KEY))
+        val currentTime = Instant.now()
+        return currentTime.minusSeconds(recTime.epochSecond).epochSecond >= TOKEN_LIFETIME
```

Comment:
What if `dataStoreManager.getString(TIME_KEY)` would return empty string, or `recTime` would happen to be after `currentTime` ?